src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
         for (src, tgt) in nodes.into_iter().tuple_windows() {
             hugr.connect(src, 0, tgt, 0)?;
         }
+        hugr.infer_and_validate(&PRELUDE_REGISTRY)?;
+        Ok(())
+    }
+
+    fn make_opaque(extension: impl Into<ExtensionId>, signature: FunctionType) -> ops::LeafOp {
+        let opaque =
+            ops::custom::OpaqueOp::new(extension.into(), "", "".into(), vec![], Some(signature));
+        ops::custom::ExternalOp::from(opaque).into()
+    }
+
+    fn make_block(
+        hugr: &mut Hugr,
+        bb_parent: Node,
+        dfb: impl Into<OpType>,
+        dfb_sig: FunctionType,
```

Comment:
I note the following in `impl Case`: https://github.com/CQCL-DEV/hugr/blob/554d658f7afd38f0eb567ceff0599e65e74541c9/src/ops/controlflow.rs#L237-L238

You could add something similar in `controlflow.rs` for `impl BasicBlock` (I guess it would have to return an Option to allow for BasicBlock::Exit), and then drop the `dfg_sig` here, making `dfb` be an `impl Into<BasicBlock>`. (More thorough still would be to turn the contents of BasicBlock::DFB into a separate struct and impl for *that*, but I'd be happy not to go as far as that in this PR)

Reply:
Or just make `make_block` here take a TypeRow for inputs and a Vec<TypeRow> for predicate, and do the rest here